src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -0,0 +1,1146 @@
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.cluster.metadata.DataStream;
+import org.opensearch.cluster.metadata.IndexAbstraction;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.unit.ByteSizeUnit;
+import org.opensearch.core.common.unit.ByteSizeValue;
+import org.opensearch.security.resolver.IndexResolverReplacer;
+import org.opensearch.security.securityconf.FlattenedActionGroups;
+import org.opensearch.security.securityconf.impl.SecurityDynamicConfiguration;
+import org.opensearch.security.securityconf.impl.v7.RoleV7;
+import org.opensearch.security.support.WildcardMatcher;
+
+import com.selectivem.collections.CheckTable;
+import com.selectivem.collections.CompactMapGroupBuilder;
+import com.selectivem.collections.DeduplicatingCompactSubSetBuilder;
+import com.selectivem.collections.ImmutableCompactSubSet;
+
+/**
+ * This class converts role configuration into pre-computed, optimized data structures for checking privileges.
+ * <p>
+ * With the exception of the statefulIndex property, instances of this class are immutable. The life-cycle of an
+ * instance of this class corresponds to the life-cycle of the role and action group configuration. If the role or
+ * action group configuration is changed, a new instance needs to be built.
+ */
+public class ActionPrivileges extends ClusterStateMetadataDependentPrivileges {
+
+    /**
+     * This setting controls the allowed heap size of the precomputed index privileges (in the inner class StatefulIndexPrivileges).
+     * If the size of the indices exceed the amount of bytes configured here, it will be truncated. Privileges evaluation will
+     * continue to work correctly, but it will be slower.
+     * <p>
+     * This settings defaults to 10 MB. This is a generous limit. Experiments have shown that an example setup with
+     * 10,000 indices and 1,000 roles requires about 1 MB of heap. 100,000 indices and 100 roles require about 9 MB of heap.
+     * (Of course, these numbers can vary widely based on the actual role configuration).
+     * <p>
+     * The setting plugins.security.privileges_evaluation.precomputed_privileges.include_indices can be used to control
+     * for which indices the precomputed privileges shall be created. This allows to lower the heap utilization.
+     */
+    public static Setting<ByteSizeValue> PRECOMPUTED_PRIVILEGES_MAX_HEAP_SIZE = Setting.memorySizeSetting(
+        "plugins.security.privileges_evaluation.precomputed_privileges.max_heap_size",
```

Comment:
To be honest, the `ConfigConstants` class grew a bit to big and hard to read for my taste. It is even not just constants, but also header values and helper methods. 

Additionally, the `ConfigConstants` follows the practice of just listing the config option names as Strings. The actual definition of the setting meta data is performed in `OpenSearchSecurityPlugin` in lines 1243 to 2048 (800 lines of just settings metadata):

https://github.com/opensearch-project/security/blob/28000a3c58f15665307ae55dafa7351e3241436f/src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java#L1247-L2050

IMHO, it would be better to keep such constants with their metadata as close to the code where they are actually being used.  That makes the code much easier to read.

Reply:
hmm, I was trying to follow suit to have similar code-style. I'm ambivalent on declaring constants in ConfigConstants file.